File: src/main/java/ludwig/workspace/SymbolRegistry.java
package ludwig.workspace;


import ludwig.changes.Delete;
import ludwig.changes.InsertNode;
import ludwig.model.*;
import ludwig.utils.NodeUtils;

import java.util.*;

import static ludwig.utils.NodeUtils.isField;

public class SymbolRegistry {
    private final TreeSet symbols = new TreeSet<>(Comparator.comparing(NodeUtils::signature));
    private final Map<String, Node> nodesById = new HashMap<>();

    public SymbolRegistry(Workspace workspace) {
        workspace.changeListeners().add(change -> {
            if (change instanceof InsertNode) {
                Node node = ((InsertNode) change).node();
                if (node instanceof FunctionNode || isField(node)) {
                    symbols.add(node);
                    nodesById.put(node.id(), node);
                }
            } else if (change instanceof Delete) {
                Node node = nodesById.remove(((Delete) change).id());
                if (node != null) {
                    deleteNode(node);
                }
            }
        });

        workspace.getProjects().forEach(this::grab);
    }

    private void deleteNode(Node<?> node) {
        symbols.remove(node);
        node.children().forEach(this::deleteNode);
    }

    public SortedSet<NamedNode> symbols(String s) {
        if (s.isEmpty() || s.length() < 2 && Character.isAlphabetic(s.charAt(0))) {
            return new TreeSet<>();
        }

        return symbols.subSet(s, s + Character.MAX_VALUE);
    }


    private void grab(Node<?> node) {
        if (node instanceof FunctionNode || isField(node)) {
            symbols.add(node);
        } else {
            node.children().forEach(this::grab);
        }

    }
}

File: src/main/kotlin/ludwig/workspace/SymbolRegistry.kt
package ludwig.workspace


import ludwig.changes.Delete
import ludwig.changes.Rename
import ludwig.model.FunctionNode
import ludwig.model.NamedNode
import ludwig.model.Node
import ludwig.model.ProjectNode
import ludwig.utils.isField
import ludwig.utils.signature
import java.util.*
import java.util.function.Consumer


class SymbolRegistry(workspace: Workspace) {
    private val symbols = TreeSet(Comparator.comparing<Any, String>({ signature(it) }))
    private val nodesById = HashMap<String, Node>()

    init {
        workspace.changeListeners().add(changeListener { change ->
            if (change is Rename) {
                val node = workspace.node(change.nodeId)
                if (node is FunctionNode || isField(node)) {
                    symbols.add(node)
                    nodesById.put(change.name, workspace.node(change.nodeId)!!)
                }
            } else if (change is Delete) {
                val node = nodesById.remove(change.nodeId)
                if (node != null) {
                    deleteNode(node)
                }
            }
        })

        workspace.projects.forEach(Consumer<ProjectNode> { this.grab(it) })
    }

    private fun deleteNode(node: Node) {
        symbols.remove(node)
        node.forEach(Consumer<Node> { this.deleteNode(it) })
    }

    fun symbols(s: String): SortedSet<NamedNode> =
            symbols.subSet(s, s + Character.MAX_VALUE) as SortedSet<NamedNode>


    private fun grab(node: Node) {
        if (node is FunctionNode || isField(node)) {
            symbols.add(node)
        } else {
            node.forEach(Consumer<Node> { this.grab(it) })
        }

    }
}
